File: dataSync/InLong-versions/inlong-1.3.0-RC1/inlong-tubemq/tubemq-client/src/main/java/org/apache/inlong/tubemq/client/consumer/ClientSubInfo.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.inlong.tubemq.client.consumer;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.TreeSet;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicBoolean;
import org.apache.inlong.tubemq.corebase.TokenConstants;

public class ClientSubInfo {
    private final ConcurrentHashMap<String/* topic */, TopicProcessor> topicCondRegistry =
            new ConcurrentHashMap<>();
    private boolean requireBound = false;
    private AtomicBoolean isNotAllocated =
            new AtomicBoolean(true);
    private int sourceCount = -1;
    private String sessionKey;
    private long subscribedTime;
    private boolean isSelectBig = true;
    private String requiredPartition = "";
    private Set<String> subscribedTopics = new HashSet<>();
    private Map<String, Long> assignedPartMap = new HashMap<>();
    private Map<String, Boolean> topicFilterMap = new HashMap<>();

    public ClientSubInfo() {

    }

    public boolean getIsNotAllocated() {
        return isNotAllocated.get();
    }

    public boolean compareAndSetIsNotAllocated(boolean expect, boolean update) {
        return this.isNotAllocated.compareAndSet(expect, update);
    }

    public boolean isSubscribedTopicEmpty() {
        return this.subscribedTopics.isEmpty();
    }

    public boolean isSubscribedTopicContain(String topic) {
        return this.subscribedTopics.contains(topic);
    }

    public TopicProcessor getTopicProcessor(String topic) {
        return this.topicCondRegistry.get(topic);
    }

    public void storeConsumeTarget(Map<String, TreeSet<String>> consumeTarget) {
        TopicProcessor topicProcessor;
        for (Map.Entry<String, TreeSet<String>> entry : consumeTarget.entrySet()) {
            topicProcessor = new TopicProcessor(null, entry.getValue());
            this.topicCondRegistry.put(entry.getKey(), topicProcessor);
            this.subscribedTopics.add(entry.getKey());
            this.topicFilterMap.put(entry.getKey(),
                    (!(entry.getValue() == null || entry.getValue().isEmpty())));
        }
        this.requireBound = false;
        this.subscribedTime = System.currentTimeMillis();
    }

    public TopicProcessor putIfAbsentTopicProcessor(String topic,
                                                    TopicProcessor topicProcessor) {
        TopicProcessor topicProcessor1 =
                this.topicCondRegistry.putIfAbsent(topic, topicProcessor);
        if (topicProcessor1 == null) {
            this.subscribedTopics.add(topic);
            Set<String> condSet = topicProcessor.getFilterConds();
            topicFilterMap.put(topic, (!(condSet == null || condSet.isEmpty())));
        }
        return topicProcessor1;
    }

    public void notifyAllMessageListenerStopped() {
        for (Map.Entry<String, TopicProcessor> entry : this.topicCondRegistry.entrySet()) {
            if (entry.getValue() != null) {
                MessageListener listener = entry.getValue().getMessageListener();
                if (listener != null) {
                    listener.stop();
                }
            }
        }
        topicCondRegistry.clear();
    }

    public int getSourceCount() {
        return sourceCount;
    }

    public String getSessionKey() {
        return sessionKey;
    }

    public void setSessionKey(String sessionKey) {
        this.sessionKey = sessionKey;
    }

    public long getSubscribedTime() {
        return subscribedTime;
    }

    public boolean isSelectBig() {
        return isSelectBig;
    }

    public boolean isRequireBound() {
        return requireBound;
    }

    public boolean isFilterConsume(String topic) {
        Boolean ret = this.topicFilterMap.get(topic);
        if (ret == null) {
            return false;
        }
        return ret;
    }

    public void setNotRequireBound() {
        this.requireBound = false;
        this.subscribedTime = System.currentTimeMillis();
    }

    /**
     * Set Bound Consumption information
     *
     * @param sessionKey     consume session key
     * @param sourceCount    the client count of consume group
     * @param isSelectBig    whether select a bigger data If there is reset conflict
     * @param partOffsetMap  the map of partitionKey and bootstrap offset
     *
     */
    public void setRequireBound(final String sessionKey,
                                final int sourceCount,
                                final boolean isSelectBig,
                                final Map<String, Long> partOffsetMap) {
        this.requireBound = true;
        this.subscribedTime = System.currentTimeMillis();
        this.sessionKey = sessionKey;
        this.isSelectBig = isSelectBig;
        this.sourceCount = sourceCount;
        int count = 0;
        StringBuilder sBuilder = new StringBuilder(256);
        for (Map.Entry<String, Long> entry : partOffsetMap.entrySet()) {
            if (entry.getKey() != null && entry.getValue() != null) {
                this.assignedPartMap.put(entry.getKey().trim(), entry.getValue());
                if (count++ > 0) {
                    sBuilder.append(TokenConstants.ARRAY_SEP);
                }
                sBuilder.append(entry.getKey().trim()).append(TokenConstants.EQ).append(entry.getValue());
            }
        }
        this.requiredPartition = sBuilder.toString();
    }

    public String getRequiredPartition() {
        return requiredPartition;
    }

    public Set<String> getSubscribedTopics() {
        return subscribedTopics;
    }

    public Map<String, Long> getAssignedPartMap() {
        return assignedPartMap;
    }

    public Long getAssignedPartOffset(String partitionKey) {
        return this.assignedPartMap.get(partitionKey);
    }

    public ConcurrentHashMap<String, TopicProcessor> getTopicCondRegistry() {
        return topicCondRegistry;
    }
}

File: dataSync/InLong-versions/inlong-1.3.0-RC1/inlong-tubemq/tubemq-client-twins/tubemq-client-cpp/src/client_subinfo.cc
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#include "client_subinfo.h"

#include "const_config.h"
#include "utils.h"



namespace tubemq {



ClientSubInfo::ClientSubInfo() {
  bound_consume_ = false;
  select_big_ = false;
  source_count_ = 0;
  session_key_ = "";
  not_allocated_.Set(true);
  is_registered_.Set(false);
  subscribed_time_ = tb_config::kInvalidValue;
  bound_partions_ = "";
}


void ClientSubInfo::SetConsumeTarget(const ConsumerConfig& config) {
  int32_t count = 0;
  string tmpstr = "";
  // book register time
  subscribed_time_ = Utils::GetCurrentTimeMillis();
  //
  is_registered_.Set(false);
  bound_consume_ = config.IsBoundConsume();
  topic_and_filter_map_ = config.GetSubTopicAndFilterMap();
  // build topic filter info
  topics_.clear();
  topic_conds_.clear();
  set<string>::iterator it_set;
  map<string, set<string> >::const_iterator it_topic;
  for (it_topic = topic_and_filter_map_.begin();
      it_topic != topic_and_filter_map_.end(); it_topic++) {
    topics_.push_back(it_topic->first);
    if (it_topic->second.empty()) {
      topic_filter_map_[it_topic->first] = false;
    } else {
      topic_filter_map_[it_topic->first] = true;

      // build topic conditions
      count = 0;
      tmpstr = it_topic->first;
      tmpstr += delimiter::kDelimiterPound;
      for (it_set = it_topic->second.begin();
          it_set != it_topic->second.end(); it_set++) {
        if (count++ > 0) {
          tmpstr += delimiter::kDelimiterComma;
        }
        tmpstr += *it_set;
      }
      topic_conds_.push_back(tmpstr);
    }
  }

  // build bound_partition info
  if (bound_consume_) {
    session_key_ = config.GetSessionKey();
    source_count_ = config.GetSourceCount();
    select_big_ = config.IsSelectBig();
    assigned_part_map_ = config.GetPartOffsetInfo();
    count = 0;
    bound_partions_ = "";
    map<string, int64_t>::const_iterator it;
    for (it = assigned_part_map_.begin();
      it != assigned_part_map_.end(); it++) {
      if (count++ > 0) {
        bound_partions_ += delimiter::kDelimiterComma;
      }
      bound_partions_ += it->first;
      bound_partions_ += delimiter::kDelimiterEqual;
      bound_partions_ += Utils::Long2str(it->second);
    }
  }
}

bool ClientSubInfo::CompAndSetNotAllocated(bool expect, bool update) {
  return not_allocated_.CompareAndSet(expect, update);
}

bool ClientSubInfo::IsFilterConsume(const string& topic) {
  map<string, bool>::iterator it;
  it = topic_filter_map_.find(topic);
  if (it == topic_filter_map_.end()) {
    return false;
  }
  return it->second;
}

void ClientSubInfo::GetAssignedPartOffset(const string& partition_key, int64_t& offset) {
  map<string, int64_t>::iterator it;
  offset = tb_config::kInvalidValue;
  if (!is_registered_.Get()
    && bound_consume_
    && not_allocated_.Get()) {
    it = assigned_part_map_.find(partition_key);
    if (it != assigned_part_map_.end()) {
      offset = it->second;
    }
  }
}

const map<string, set<string> >& ClientSubInfo::GetTopicFilterMap() const {
  return topic_and_filter_map_;
}


}  // namespace tubemq

